### PR TITLE
lora to lm_head

### PR DESCRIPTION
**Pull request by jamesascarter**

### CODE REVIEW

File: model/modules/custom_lora.py [hunk 1]
```diff
@@ -6,3 +6,35 @@
...
+        self.A = nn.Linear(config.input_dim, config.rank, bias=False)
+        self.B = nn.Linear(config.rank, config.output_dim, bias=True)
+        
+    def forward(self, x):
+        # x: (batch_size, seq_len, input_dim)
+        x = self.A(x)  # (batch_size, seq_len, rank)
+        x = F.relu(x)
+        x = self.dropout(x)
+        x = self.B(x)  # (batch_size, seq_len, output_dim)
+        return x
```
Comment (dhedey) [inline model/modules/custom_lora.py]: Reading up on LoRA, it's typically done with:

return lora_B(lora_A(dropout(x))) * self.alpha

And no RELU

File: model/modules/qwen_multi_modal_model.py [hunk 2]
```diff
@@ -152,6 +154,16 @@ def __init__(self, config: QwenMultiModalModelConfig):
...
         ))
 
+        if config.apply_lora_to_lm_head_layer:
+            self.lm_head_lora_diff = CustomLora(CustomLoraConfig(
+                input_dim=self.embedding_dimension,
+                output_dim=len(self.tokenizer),  # vocab size
+                rank=64,  # or make it configurable
+                dropout=0.1
+            ))
+        else:
```
Comment (dhedey) [inline model/modules/qwen_multi_modal_model.py]: Arr sorry there might be some merge conflicts here!

You can put it under here now:
```
            case EmbeddingLearningStrategy.LORA:
                raise NotImplementedError("LoRA has not been implemented for lm_head yet")
```

**Review by dhedey — APPROVED** [review]
